File: cases/taint/java/nested-access-path-negative/NestedAccessPathNegative.java
package dataflowbench.taint;

final class NestedAccessPathNegative {
    static final class Leaf {
        int value;
        int other;
    }

    static final class Middle {
        final Leaf c = new Leaf();
    }

    static final class Outer {
        final Middle b = new Middle();
    }

    static int dfb_source() { // DFB-SOURCE: nested-access-path-input
        return 1;
    }

    static void dfb_sink(int value) { } // DFB-SINK: nested-access-path-sink

    static void run() {
        Outer a = new Outer();
        a.b.c.value = dfb_source(); // DFB-WITNESS: nested-access-path-store
        a.b.c.other = 0;
        dfb_sink(a.b.c.other);
    }
}

File: cases/taint/kotlin/nested-access-path-negative/NestedAccessPathNegative.kt
package dataflowbench

object NestedAccessPathNegative {
    class Inner {
        var value: String = "clean"
        var other: String = "clean"
    }

    class Middle {
        val inner: Inner = Inner()
    }

    class Outer {
        val middle: Middle = Middle()
    }

    fun dfb_source(): String { // DFB-SOURCE: nested-access-path-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: nested-access-path-sink

    fun run() {
        val outer = Outer()
        outer.middle.inner.value = dfb_source() // DFB-WITNESS: nested-access-path-store
        outer.middle.inner.other = "clean"
        dfb_sink(outer.middle.inner.other)
    }
}
